File: src/main/java/eu/karelhovorka/zpevnik/text/SectionType.java
package eu.karelhovorka.zpevnik.text;


import java.util.Locale;

import eu.karelhovorka.zpevnik.util.Preconditions;

public enum SectionType {
    INTRO("I", "I:", "Intro:"),
    CHORUS("R", "R([0-9]*):", "Chorus:", "Refren:"),
    VERSE("S", "S([0-9]*):", "Sloka:", "V:", "([0-9]+)\\."),
    BRIDGE("B", "B:", "Bridge:"),
    NOTE("P", "Note:", "Poznamka:"),
    SOLO("Solo", "Solo:"),
    INTERMEZZO("M", "Mezihra:"),
    OUTRO("O", "Outro:"),
    SPEECH("Rec", "Recital:", "Rec:"),
    UNKNOWN("?", "");

    public final String names[];
    private final String shortcut;

    public String css() {
        return toString().toLowerCase(Locale.FRANCE);
    }

    SectionType(String shortcut, String... names) {
        Preconditions.checkArgument(names.length > 0, "at least 1 name must exist");
        this.names = names;
        this.shortcut = shortcut;
    }

    public static SectionType fromName(String line) {
        for (SectionType sectionType : values()) {
            for (String name : sectionType.names) {
                if (line.matches(name)) {
                    return sectionType;
                }
            }
        }
        //return UNKNOWN;
        throw new IllegalArgumentException("line: " + line);
    }

    public String getShortcut(int index) {
        if (index == 0) {
            return shortcut;
        }
        return shortcut + (index + 1);
    }
}

File: src/main/java/eu/karelhovorka/zpevnik/text/SectionType.kt
package eu.karelhovorka.zpevnik.text


import eu.karelhovorka.zpevnik.util.Preconditions
import java.util.*

interface ISectionType {
    fun css(): String
    fun getShortcut(index: Int): String
    fun getLongName(index: Int): String
}

class UnknownSectionType(val name: String): ISectionType {
    override fun css(): String {
        return "unknown-section"
    }

    override fun getShortcut(index: Int): String {
        return if (index == 0) {
            name
        } else  "$name(${index + 1})"
    }

    override fun getLongName(index: Int): String {
        return if (index == 0) {
            name
        } else  "$name(${index + 1})"
    }

}

enum class SectionType private constructor(private val shortcut: String, private val longName: String, vararg names: String) : ISectionType {
    INTRO("I", "Předehra", "I:", "U:", "Intro:", "Předehra:", "Úvod"),
    CHORUS("R", "Refrén", "R([0-9]*):", "Chorus:", "Refren:", "Ref:"),
    VERSE("S", "Sloka", "S([0-9]*):", "Sloka:", "V:", "([0-9]+)\\."),
    BRIDGE("B", "Bridge", "Bridge:", "B:"),
    NOTE("P", "Poznámka", "Note:", "Poznamka:"),
    SOLO("Solo", "Solo", "Solo:"),
    INTERMEZZO("M", "Mezihra", "Mezihra:", "M:"),
    OUTRO("O", "Outro", "Outro:"),
    SPEECH("Rec", "Recitál", "Recital:", "Rec:"),
    UNKNOWN("?", "?", "");

    val names: Array<String>

    override fun css(): String {
        return toString().toLowerCase(Locale.FRANCE)
    }

    init {
        Preconditions.checkArgument(names.size > 0, "at least 1 name must exist")
        this.names = names as Array<String>
    }

    override fun getShortcut(index: Int): String {
        return if (index == 0) {
            shortcut
        } else "$shortcut(${index + 1})"
    }

    override fun getLongName(index: Int): String {
        return if (index == 0) {
            longName
        } else  "$longName(${index + 1})"

    }

    companion object {

        fun fromName(line: String): ISectionType {
            for (sectionType in values()) {
                for (name in sectionType.names) {
                    if (line.trim().matches(name.toRegex())) {
                        return sectionType
                    }
                }
            }
            return UnknownSectionType(line.trim())
        }
    }
}
